File: Factory Method/factoryMethod.ts
abstract class PerfilUsuario {
  nome: string;

  constructor(nome: string) {
    this.nome = nome;
  }
}

class Convidado extends PerfilUsuario {
  tipo_perfil: string;

  constructor(nome: string) {
    super(nome);
    this.tipo_perfil = 'Convidado';
  }
}

class PessoaJuridica extends PerfilUsuario {
  tipo_perfil: string;

  constructor(nome: string) {
    super(nome);
    this.tipo_perfil = 'Pessoa Jurídica';
  }
}

class PessoaFisica extends PerfilUsuario {
  tipo_perfil: string;

  constructor(nome: string) {
    super(nome);
    this.tipo_perfil = 'Pessoa Física';
  }
}

interface FabricaUsuario {
  criar_perfil(nome: string): PerfilUsuario;
}

class Usuario implements FabricaUsuario {
  nome: string;

  constructor(nome: string) {
    this.nome = nome;
  }

  criar_perfil(tipo_perfil: string) {
    if (tipo_perfil === 'convidado') {
      return new Convidado(this.nome);
    } else if (tipo_perfil === 'pessoa_juridica') {
      return new PessoaJuridica(this.nome);
    } else if (tipo_perfil === 'pessoa_fisica') {
      return new PessoaFisica(this.nome);
    } else {
      throw new Error('Tipo de perfil inválido');
    }
  }
}

// USO

const nome_usuario = 'Jardel';
const usuario = new Usuario(nome_usuario);

const novoPerfil = usuario.criar_perfil('pessoa_fisica');

console.log(
  `\nPerfil criado:  ${novoPerfil.tipo_perfil} - ${novoPerfil.nome}\n`
);

File: Factory Method/factoryMethod.py
from abc import ABC, abstractmethod
from enum import Enum


class PerfilUsuario(ABC):
    def __init__(self, nome):
        self.nome = nome


class Convidado(PerfilUsuario):
    def __init__(self, nome):
        super().__init__(nome)
        self.tipo_perfil = "Convidado"


class PessoaJuridica(PerfilUsuario):
    def __init__(self, nome):
        super().__init__(nome)
        self.tipo_perfil = "Pessoa Jurídica"


class PessoaFisica(PerfilUsuario):
    def __init__(self, nome):
        super().__init__(nome)
        self.tipo_perfil = "Pessoa Física"


class FabricaUsuario(ABC):
    @abstractmethod
    def criar_perfil(self, nome):
        pass


class Usuario(FabricaUsuario):
    def __init__(self, nome):
        self.nome = nome

    def criar_perfil(self, tipo_perfil):
        if tipo_perfil == "convidado":
            return Convidado(self.nome)
        elif tipo_perfil == "pessoa_juridica":
            return PessoaJuridica(self.nome)
        elif tipo_perfil == "pessoa_fisica":
            return PessoaFisica(self.nome)
        else:
            raise ValueError("Tipo de perfil inválido")


# USANDO


nome_usuario = input("Digite seu nome: ")
usuario = Usuario(nome_usuario)

print("\nEscolha o tipo de perfil que deseja criar:")
print("1. Convidado")
print("2. Pessoa Jurídica")
print("3. Pessoa Física\n")

opcao = int(input("Digite a opção desejada: "))

if opcao == 1:
    perfil = usuario.criar_perfil("convidado")
elif opcao == 2:
    perfil = usuario.criar_perfil("pessoa_juridica")
elif opcao == 3:
    perfil = usuario.criar_perfil("pessoa_fisica")
else:
    print("Opção inválida")
    exit()


if perfil:
    print(f"\nPerfil criado: {perfil.tipo_perfil} - {perfil.nome}\n")
